tracer/src/Datadog.Trace/Agent/MessagePack/SpanMessagePackFormatter.cs

```diff
@@ -275,14 +298,30 @@ private int WriteMetrics(ref byte[] bytes, int offset, in SpanModel model, ITagP
             offset = tagWriter.Offset;
             count += tagWriter.Count;
 
-            if (model.IsLocalRoot)
+            // add "process_id" tag to local root span (if present)
+            if (model.IsLocalRoot && _processIdValueBytes != null)
             {
-                // add "process_id" tag
-                if (_processIdValueBytes != null)
+                count++;
+                offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, _processIdNameBytes);
+                offset += MessagePackBinary.WriteRaw(ref bytes, offset, _processIdValueBytes);
+            }
+
+            // add "_sampling_priority_v1" tag to all "chunk orphans"
+            // (spans whose parents are not found in the same chunk)
+            if (model.IsChunkOrphan && model.TraceChunk.SamplingPriority is { } samplingPriority)
```

Comment:
> [agent versions >= 7.34.0] iterate the entire trace chunk the first span it find with the sampling priority tag, it uses that one

Should we also add the sampling prio on the first span to avoid degrading performances on the agent side?

Reply:
That's a good catch, but I don't think we need it because the agent checks for a trace chunk's sampling priority in the following order:

1. if any span has `parentId = 0`
    - in this case we add the sampling priority to this span only (and we stop looking for other "chunk orphans")
2. out of all the "chunk orphans", it picks one (non-deterministically)
    - in this case we add the sampling priority to all eligible spans
    - even if we are not partial flushing, this can happen if there is an upstream service because no span has `parentId = 0`
3. it iterates all spans from the start of the trace chunks
    - the agent should never fall into this case case. for this to happen, there would need to be a circular reference in the tree structure (no chunk orphans), and that would not be valid trace chunk.

Reply:
I updated the PR description to try to make the agent steps more clear, and to specify what tracers can do to help the agent.

Reply:
I thought about this some more and I want to write more thoughts down because I will forget later. If we detect agent version >= 7.34.0, in theory it is enough to add the sampling priority to the first span of every chunk.

However, this is not faster for the agent (they will still try 1 and 2 above before trying 3), so it's trading work in the tracer for work in the agent. In addition, this fallback does not work for the `_dd.origin` tag, which will keep looking at the "chunk root" span.

So to be as correct as possible, I think we should find the chunk orphans even for newer versions of the agent. Also, maybe the agent team can refactor all that agent code to make it much more efficient.